### PR TITLE
support multiple local servers in configuration file

### PR DESCRIPTION
**Pull request by zonyitoo**
ref #452

- support `locals` in configuration file, running multiple local server
  instance simultaneously
- support `unix://` in `dns` configuration

BREAKING CHANGE:

- `sslocal`'s `--dns-addr` is now only available in Android
- shadowsocks-service's `Config` struct have lots of changes

### CODE REVIEW

**Comment by zonyitoo** [comment]
Invite @madeye for testing interpolation into shadowsocks-android.

**Comment by madeye** [comment]
I will take a look later.

**Review by database64128 — APPROVED** [review]

**Comment by madeye** [comment]
Anything changed to `local_address` and `local_port` field?

I see this error:

```
03-13 19:48:50.288 23560 23972 E libsslocal: error: The following required arguments were not provided:
03-13 19:48:50.289 23560 23972 E libsslocal:     --local-addr <LOCAL_ADDR>
```

**Comment by zonyitoo** [comment]
Oh, that's because `-U` and `-u` are now binded to the `local-addr` parameter, which means to set `mode` for that specific local instance.

So please just remove that `-U` or `-u` from the command line argument and place it into the configuration file.

**Comment by zonyitoo** [comment]
```
--local-addr xx --protocol xx --forward-addr xx --local-dns-addr xx --remote-dns-addr xx -U -v
```

These arguments are now changed for creating ONE new local instance.

File: bin/ssmanager.rs [hunk 1]
```diff
@@ -50,7 +47,7 @@ fn main() {
                 the only required fields are \"manager_address\" and \"manager_port\". \
                 Servers defined will be created when process is started.")
 
-        (@arg BIND_ADDR: -b --("bind-addr") +takes_value "Bind address, outbound socket will bind this address")
+        (@arg BIND_ADDR: -b --("bind-addr") +takes_value {validator::validate_ip_addr} "Bind address, outbound socket will bind this address")
```
Comment (tony) [inline bin/ssmanager.rs:50]: Looks good.

**Comment by madeye** [comment]
Okay, it works now.

LGTM

**Comment by zonyitoo** [comment]
Great. Merging v1.10.0 now.

**Comment by zonyitoo** [comment]
Oh, one more thing, @madeye , you should set `dns` to `"dns": "unix://xxxx` for connecting Android's UDS DNS server.

**Comment by madeye** [comment]
I'm still using `local-dns-addr` arg for the local DNS resolver.

**Comment by zonyitoo** [comment]
Yes it should work, but `sslocal`'s own DNS queries will still need the `dns` arg. Maybe I should expose it to command line argument?

**Comment by madeye** [comment]
Okay, let me try the `dns` filed in config file first.

We're still passing IP address to ss-rust, that's why I didn't notice any issue before.

**Comment by zonyitoo** [comment]
With this new locals format, you can start it like

```cjson
{
    "locals": [
        // Normal socks5 server
        { "local_address": 127.0.0.1, "local_port": 1081, "local_udp_address": "127.0.0.1", "local_udp_port": 2081, "mode": "tcp_and_udp" },
        // DNS server
        { "local_address": 127.0.0.1, "local_port": 53, "local_dns_address": "unix://yourpath", "remote_dns_address": "8.8.8.8" }
    ]
}
```

**Comment by madeye** [comment]
Yes, it works well now.

**Comment by zonyitoo** [comment]
You can delete most of the command line arguments, except `--stat-path`.

**Pull request merged**